### DISCUSSION TITLE
Use NLP for Action Routing?

### DISCUSSION THREAD

**Discussion by judahpaul16** [community/community#125994] [Programming Help] [answered]
### Body

I'm considering switching from regex to NLP models for action routing and I'm curious about others' experiences with this. The main draw for NLP is its ability to understand context which allows for more natural interactions. It doesn't need constant updates like regex when new intents or variations arise, and it's generally better at handling ambiguous queries.

However, there are notable downsides. NLP models can be resource-heavy, potentially slowing down interactions due to increased computational demands. The integration process is more complex than with regex, involving additional steps like managing the model's lifecycle. Also, maintaining performance might require ongoing training and tuning, which adds a layer of overhead that you don't get with static regex patterns.

Does anyone have any ideas or experience with using NLP for routing control flow?

```python
async def action_router(text: str, display):
    # Alarm and Reminder actions
    if re.search(r'\b(alarm|timer|reminder|remind me|^wake me up)\b', text, re.IGNORECASE):
        return await alarm_reminder_action(text)
    
    # For Spotify actions
    if re.search(r'\b(^play|resume|next song|go back|pause|stop|shuffle|repeat|volume)(\s.*)?(\bon\b\sSpotify)?\b', text, re.IGNORECASE):
        return await spotify_action(text)
        
    # For Open Weather actions
    elif re.search(r'\b(weather|forecast|temperature)\b.*\b(in|for|at)?\b(\w+)?', text, re.IGNORECASE) or \
        re.search(r'\b(is|will)\sit\b.*\b(hot|cold|rain(ing|y)?|sun(ny|ning)?|cloud(y|ing)?|wind(y|ing)?|storm(y|ing)?|snow(ing)?)\b', text, re.IGNORECASE):
        return await open_weather_action(text)

    # For Philips Hue actions
    elif re.search(r'\b(turn\s)?(the\s)?lights?\s?(on|off)?\b', text, re.IGNORECASE):
        return await philips_hue_action(text)

    # For Calendar & To-Do List actions (CalDAV)
    elif re.search(r'\b(event|calendar|schedule|appointment|task|todo|to-do|task\slist|to-do\slist|to do)\b', text, re.IGNORECASE):
        return await caldav_action(text)

    # If no pattern matches, query OpenAI
    else:
        return await query_openai(text, display)
```

Original [context](https://github.com/judahpaul16/gpt-home/discussions/48).

### Guidelines

- [X] I have read and understood this category's [guidelines](https://github.com/orgs/community/discussions/51384) before making this post.

**Comment by CopperEagle** [comment, marked as answer]
Hey @judahpaul16 

The main problem you will deal with is latency. Now if you use ChatGPT-4o, you may be fine with adding an NLP step. 

## Interface

So first off, here is my recommendation: Whatever NLP method you use, abstract it behind some simple interface (using a class). This way, you can switch the implementation of the class (the NLP method) easily without changing anything else. A possible interface looks like this

```py
class ActionRouter:
    def initialize(self, *args, **kwargs): ...

    def shutdown(self, *args, **kwargs): ...

    def resolve(self, text): ...
        """This thing returns Action objects."""

    def get_method_name(self): ...

class Action:
    def __init__(self, type, text): ...
        """Type: alarm, spotify, etc."""

    async def perform(self, **kwarg): ...
        """Case distinction between different types, then do the action."""
```

This is quite flexible for future development including action logging, etc. Your `action_router` function now looks a lot easier. Note that you can still use `alarm_reminder_action`, etc. as is inside `Action.perform`. The `**kwarg` ensures that if you pass on display, the variable will be named accordingly, see below:


```py
async def action_router(text : str, display):
    act = router.resolve(text)
    act.perform(display=display)

# elsewhere
import atexit
# ...

router = ActionRouter(...)
router.initialize(...)
atexit.register(router.shutdown)
```

### Implementation of the ActionRouter class

This is the choice of the NLP method. You can really go wild here. For me, it is hard to give you recommendations here as I don't know how much hardware you have available and what latency you tolerate.

On one end you can use Huggingface and its `transformers` library. There you could use even some LLM like `LLama-3-8B`. If you want a smaller model, there are some T5 derivatives that are quite good and are only ~3 GB in size. This should be faster, too. The other end are more traditional NLP methods. It depends on your tolerances for latency and error and hardware. [HERE](https://medium.com/@adampaslawski/llm-routing-with-semantics-taking-the-right-actions-0fda470c1712) is an article with a morer thorough discussion of action routing implementations.

Here is an example for a T5 derivative

```py
from transformers import AutoTokenizer, AutoModelForSeq2SeqLM
from transformers import pipeline

##### DOWNLOADING/LOADING FROM MEMORY
tokenizer = AutoTokenizer.from_pretrained("MBZUAI/LaMini-Flan-T5-783M")
tokenizer.save_pretrained("./tokenizer")

model = AutoModelForSeq2SeqLM.from_pretrained("MBZUAI/LaMini-Flan-T5-783M", revision="refs/pr/6", use_safetensors=True)
model.save_pretrained("./model", max_shard_size="200MB")

##### INITIALIZATION
model = pipeline(task= 'text2text-generation', model=model, tokenizer=tokenizer)

##### GET ANSWER
response = model("Hey, how are you?", max_length=2048, do_sample=True)[0]['generated_text']
```

Of course, you then need to turn a model answer into an action. Again, check the article I linked before.

### Stay with regex?

If latency matters to you, then you should maybe try to not use NLP for action routing. A possible way to scale this better is to *build* regex strings at startup time e.g. like this:

```py
KW_ALARM = ["alarm","timer","reminder","remind me","^wake me up"]
REG_ALARM = r"\b(" + r"|".join(KW_ALARM) + r"\b"
```

**Reply by CopperEagle** [reply]
Sorry for wall of text :sweat_smile:
